### DISCUSSION TITLE
Circuit Box corrupted a multiplayer save file

### DISCUSSION THREAD

**Discussion by CianIrish** [FakeFishGames/Barotrauma#13904] [Bug reports]
#13904 ## Disclaimers

- [X] I have searched the issue tracker to check if the issue has already been reported.
- [ ] My issue happened while using mods.

### What happened?

I was messing around with the circuit box, there were many different components in it. 

The circuit box already had a reactor controller working, when we went to undock from the outpost the game crashed and everyone was unable to load the save back up.

Sorry If the information included is not useful, please ask away and I will provide anything needed

*Edit: It looks like another player has had the same issue, the issue seems to be from messing around with labels in the circuitbox.

Save file attached
[Blood_in_The_Water_Update.zip](https://github.com/FakeFishGames/Barotrauma/files/15156916/Blood_in_The_Water_Update.zip)

### Reproduction steps

Unable to reproduce as I done a lot of testing in the circuit box before we undocked

### Bug prevalence

Happens every time I play

### Single player or multiplayer?

Multiplayer hosted from the in-game menu (= using a listen server)

### Version

v1.4.5.0 (Blood in the Water Update, hotfix 1)

### Which operating system did you encounter this bug on?

Windows

### Relevant error messages and crash reports

```shell
Barotrauma Client crash report (generated on 04/29/2024 21:00:43)

Barotrauma seems to have crashed. Sorry for the inconvenience! 

39F8D09D85D1DF9EA438EE2146095747

Game version 1.4.5.0 (ReleaseWindows, branch release/spring-update-hotfix, revision 5ad6212551)
Graphics mode: 1920x1080 (BorderlessWindowed)
VSync ON
Language: English
Selected content packages: Vanilla (E218DD5)
Level seed: no level loaded
Loaded submarine: None
Selected screen: Barotrauma.RoundSummaryScreen
SteamManager initialized
Client (Round hadn't been started)

System info:
    Operating system: Microsoft Windows NT 10.0.19045.0 64 bit
    GPU name: NVIDIA GeForce RTX 4070 Ti
    Display mode: {Width:1920 Height:1080 Format:Color AspectRatio:1.7777778}
    GPU status: Normal

Exception: Length cannot be less than zero. (Parameter 'length') (System.ArgumentOutOfRangeException)
Target site: System.String Substring(Int32, Int32)
Stack trace: 
   at System.String.Substring(Int32 startIndex, Int32 length)
   at Barotrauma.ToolBox.LimitStringHeight(String str, ScalableFont font, Int32 maxHeight) in <DEV>\Barotrauma\BarotraumaClient\ClientSource\Utils\ToolBox.cs:line 444
   at Barotrauma.CircuitBoxLabelNode.UpdateTextSizes(RectangleF rect) in <DEV>\Barotrauma\BarotraumaClient\ClientSource\CircuitBox\CircuitBoxLabelNode.cs:line 40
   at Barotrauma.CircuitBoxLabelNode.LoadFromXML(ContentXElement element, CircuitBox circuitBox) in <DEV>\Barotrauma\BarotraumaShared\SharedSource\CircuitBox\CircuitBoxLabelNode.cs:line 80
   at Barotrauma.Items.Components.CircuitBox.LoadFromXML(ContentXElement loadElement) in <DEV>\Barotrauma\BarotraumaShared\SharedSource\Items\Components\Signal\CircuitBox.cs:line 185
   at Barotrauma.Items.Components.CircuitBox.TryInitializeNodes() in <DEV>\Barotrauma\BarotraumaShared\SharedSource\Items\Components\Signal\CircuitBox.cs:line 163
   at Barotrauma.Items.Components.CircuitBox.OnMapLoaded() in <DEV>\Barotrauma\BarotraumaShared\SharedSource\Items\Components\Signal\CircuitBox.cs:line 156
   at Barotrauma.Item.OnMapLoaded() in <DEV>\Barotrauma\BarotraumaShared\SharedSource\Items\Item.cs:line 4165
   at Barotrauma.MapEntity.MapLoaded(List`1 entities, Boolean updateHulls) in <DEV>\Barotrauma\BarotraumaShared\SharedSource\Map\MapEntity.cs:line 825
   at Barotrauma.Submarine..ctor(SubmarineInfo info, Boolean showErrorMessages, Func`2 loadEntities, IdRemap linkedRemap) in <DEV>\Barotrauma\BarotraumaShared\SharedSource\Map\Submarine.cs:line 1684
   at Barotrauma.GameSession.StartRound(LevelData levelData, Boolean mirrorLevel, SubmarineInfo startOutpost, SubmarineInfo endOutpost) in <DEV>\Barotrauma\BarotraumaShared\SharedSource\GameSession\GameSession.cs:line 451
   at Barotrauma.CampaignMode.<>c__DisplayClass36_0.<<SelectSummaryScreen>b__0>d.MoveNext() in <DEV>\Barotrauma\BarotraumaClient\ClientSource\GameSession\GameModes\CampaignMode.cs:line 314
--- End of stack trace from previous location ---
   at Barotrauma.GameMain.Update(GameTime gameTime) in <DEV>\Barotrauma\BarotraumaClient\ClientSource\GameMain.cs:line 942
   at Microsoft.Xna.Framework.Game.DoUpdate(GameTime gameTime) in <DEV>\Libraries\MonoGame.Framework\Src\MonoGame.Framework\Game.cs:line 656
   at Microsoft.Xna.Framework.Game.Tick() in <DEV>\Libraries\MonoGame.Framework\Src\MonoGame.Framework\Game.cs:line 500
   at Microsoft.Xna.Framework.SdlGamePlatform.RunLoop() in <DEV>\Libraries\MonoGame.Framework\Src\MonoGame.Framework\SDL\SDLGamePlatform.cs:line 93
   at Microsoft.Xna.Framework.Game.Run(GameRunBehavior runBehavior) in <DEV>\Libraries\MonoGame.Framework\Src\MonoGame.Framework\Game.cs:line 397
   at Microsoft.Xna.Framework.Game.Run() in <DEV>\Libraries\MonoGame.Framework\Src\MonoGame.Framework\Game.cs:line 367
   at Barotrauma.Program.Main(String[] args) in <DEV>\Barotrauma\BarotraumaClient\ClientSource\Program.cs:line 61

Last debug messages:
[04/29/2024 21:00:37] Result "UnexpectedError" was not handled by "UnlockAchievements".
[04/29/2024 20:33:07] Result "UnexpectedError" was not handled by "UnlockAchievements".
[04/29/2024 20:33:06] Received STARTGAMEFINALIZE packet. Round init status: Started
[04/29/2024 20:33:06] Received STARTGAMEFINALIZE packet. Round init status: WaitingForStartGameFinalize
[04/29/2024 20:32:56] Generating an outpost for the start of the level... (Location: Xinzou, level type: Outpost)
[04/29/2024 20:32:56] Level identifier: ridgeoutpost
[04/29/2024 20:32:41] Result "UnexpectedError" was not handled by "UnlockAchievements".
[04/29/2024 20:32:41] Result "UnexpectedError" was not handled by "UnlockAchievements".
[04/29/2024 20:32:41] Result "UnexpectedError" was not handled by "UnlockAchievements".
[04/29/2024 20:32:41] Result "UnexpectedError" was not handled by "UnlockAchievements".
[04/29/2024 20:32:41] Result "UnexpectedError" was not handled by "UnlockAchievements".
[04/29/2024 20:32:41] Result "UnexpectedError" was not handled by "UnlockAchievements".
[04/29/2024 20:32:41] Result "UnexpectedError" was not handled by "UnlockAchievements".
[04/29/2024 20:29:46] Result "UnexpectedError" was not handled by "UnlockAchievements".
[04/29/2024 20:29:30] WARNING: Received a potentially invalid ENTITY_POSITION message.Entity identifier does not match (server entity is spinelingspike, client entity is chaingunboltshredder). Ignoring the message...
[04/29/2024 20:25:00] WARNING: Received a potentially invalid ENTITY_POSITION message.Entity identifier does not match (server entity is chaingunboltshredder, client entity is coilgunboltpiercing). Ignoring the message...
[04/29/2024 20:25:00] WARNING: Received a potentially invalid ENTITY_POSITION message.Entity identifier does not match (server entity is chaingunboltshredder, client entity is coilgunboltpiercing). Ignoring the message...
[04/29/2024 20:24:59] WARNING: Received a potentially invalid ENTITY_POSITION message.Entity identifier does not match (server entity is chaingunboltshredder, client entity is coilgunboltpiercing). Ignoring the message...
[04/29/2024 20:24:19] Result "UnexpectedError" was not handled by "UnlockAchievements".
[04/29/2024 20:21:47] Result "UnexpectedError" was not handled by "UnlockAchievements".
[04/29/2024 20:20:27] Result "UnexpectedError" was not handled by "UnlockAchievements".
[04/29/2024 20:20:09] Result "UnexpectedError" was not handled by "UnlockAchievements".
[04/29/2024 20:19:53] Result "UnexpectedError" was not handled by "UnlockAchievements".
[04/29/2024 20:18:46] Result "UnexpectedError" was not handled by "UnlockAchievements".
[04/29/2024 20:16:59] WARNING: Received a potentially invalid ENTITY_POSITION message.Entity identifier does not match (server entity is chaingunboltshredder, client entity is coilgunboltpiercing). Ignoring the message...
[04/29/2024 20:16:58] WARNING: Received a potentially invalid ENTITY_POSITION message.Entity identifier does not match (server entity is chaingunboltshredder, client entity is coilgunboltpiercing). Ignoring the message...
[04/29/2024 20:14:43] Result "UnexpectedError" was not handled by "UnlockAchievements".
[04/29/2024 20:14:41] Received STARTGAMEFINALIZE packet. Round init status: Started
[04/29/2024 20:14:41] Received STARTGAMEFINALIZE packet. Round init status: Started
[04/29/2024 20:14:41] Received STARTGAMEFINALIZE packet. Round init status: Started
[04/29/2024 20:14:41] Received STARTGAMEFINALIZE packet. Round init status: Started
[04/29/2024 20:14:40] Received STARTGAMEFINALIZE packet. Round init status: WaitingForStartGameFinalize
[04/29/2024 20:14:33] Generating an outpost for the end of the level... (Location: Xinzou, level type: LocationConnection)
[04/29/2024 20:14:32] Generating an outpost for the start of the level... (Location: Newellstead Fortification, level type: LocationConnection)
[04/29/2024 20:14:32] Level identifier: ridgebasic
[04/29/2024 20:11:56] Result "UnexpectedError" was not handled by "UnlockAchievements".
[04/29/2024 20:11:55] Received STARTGAMEFINALIZE packet. Round init status: Started
[04/29/2024 20:11:54] Received STARTGAMEFINALIZE packet. Round init status: WaitingForStartGameFinalize
[04/29/2024 20:11:52] Generating an outpost for the end of the level... (Location: Xinzou, level type: LocationConnection)
[04/29/2024 20:11:50] Generating an outpost for the start of the level... (Location: Newellstead Fortification, level type: LocationConnection)
[04/29/2024 20:11:49] Level identifier: ridgebasic
[04/29/2024 18:56:31] Result "UnexpectedError" was not handled by "UnlockAchievements".
[04/29/2024 18:56:30] Received STARTGAMEFINALIZE packet. Round init status: Started
[04/29/2024 18:56:30] Received STARTGAMEFINALIZE packet. Round init status: WaitingForStartGameFinalize
[04/29/2024 18:56:21] Generating an outpost for the start of the level... (Location: Newellstead Fortification, level type: Outpost)
[04/29/2024 18:56:21] Level identifier: ridgeoutpost
[04/29/2024 18:55:04] Result "UnexpectedError" was not handled by "UnlockAchievements".
[04/29/2024 18:55:02] Received STARTGAMEFINALIZE packet. Round init status: Started
[04/29/2024 18:55:02] Received STARTGAMEFINALIZE packet. Round init status: Started
[04/29/2024 18:55:02] Received STARTGAMEFINALIZE packet. Round init status: Started
[04/29/2024 18:55:02] Received STARTGAMEFINALIZE packet. Round init status: Started
[04/29/2024 18:55:01] Received STARTGAMEFINALIZE packet. Round init status: WaitingForStartGameFinalize
[04/29/2024 18:54:53] Generating an outpost for the end of the level... (Location: Xinzou, level type: LocationConnection)
[04/29/2024 18:54:53] Generating an outpost for the start of the level... (Location: Newellstead Fortification, level type: LocationConnection)
[04/29/2024 18:54:53] Level identifier: ridgebasic
[04/29/2024 18:54:12] Received STARTGAMEFINALIZE packet. Round init status: Started
[04/29/2024 18:54:12] Received STARTGAMEFINALIZE packet. Round init status: Started
[04/29/2024 18:54:12] Received STARTGAMEFINALIZE packet. Round init status: Started
[04/29/2024 18:54:12] Received STARTGAMEFINALIZE packet. Round init status: Started
[04/29/2024 18:54:12] Received STARTGAMEFINALIZE packet. Round init status: WaitingForStartGameFinalize
[04/29/2024 18:54:02] Generating an outpost for the end of the level... (Location: Bramerton Outpost, level type: LocationConnection)
[04/29/2024 18:54:01] Level identifier: wastesbasic
```

**Comment by ColColonCleaner** [comment]
Running into a similar issue here. Could be the same cause.

**Comment by 3e849f2e5c** [comment]
This should be fixed in the hotfix coming out soon

**Reply by CianIrish** [reply]
I'd say so, I fixed my save but only by editing the subs xml.

In my case, I had written in the label and skipped a load of lines and then written again, when looking at the xml it showed something like text="Turbine Output  &#xA&#xA&#xA&#xA&#xA&#xA&#xA&#xA&#xA&#xA&#xA&#xA&#xA&#xA&#xA Fission Rate"

So to me, it looks like when that string is passed to ToolBox.LimitStringHeight doesn't see it as just one long string so when it goes to pass on information to System.String.Substring(Int32 startIndex, Int32 length) but it's not given what is expected and is out of range.

If I could see what the output was for ToolBox.LimitStringHeight(String str, ScalableFont font, Int32 maxHeight) then I could confirm if System.String.Substring(Int32 startIndex, Int32 length) is receiving the correct values.

TL;DR Fixed save by removing line breaks in label text in subs xml.

**Reply by 3e849f2e5c** [reply]
Closing discussion, see https://github.com/FakeFishGames/Barotrauma/discussions/13905#discussioncomment-9293491
